Jetpack Manage: Implement a modal for the "Get Score" action for Boost

yashwin opened the pull request:
Resolves https://github.com/Automattic/jetpack-genesis/issues/26

## Proposed Changes

- This PR implements a modal for "Get Score" action for Boost

### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/implement-get-score-modal` and `yarn start-jetpack-cloud`.
2. Open http://jetpack.cloud.localhost:3000/, and you'll be redirected to the /dashboard.
3. 3. Apply this patch: D123108-code
4. Add a new JN Site > Click the `Get Score` button as shown below

<img width="197" alt="Screenshot 2023-09-27 at 5 16 35 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/a80f1587-6053-4685-91e8-62add6a20420">

5. Verify that you can now see the modal as shown below with the changes in CTAs.

<img width="1065" alt="Screenshot 2023-09-28 at 10 03 52 AM" src="https://github.com/Automattic/wp-calypso/assets/10586875/30f26421-0206-4784-ab48-d1c09f474113">

6. Click the Jetpack Boost link and verify it works and redirect to https://wordpress.org/plugins/jetpack-boost/

7. Click the `Purchase Boost License` and verify you are being redirected to the add license page.

8. Repeat step 3 and, click the `Start Free` button, and verify that we are making an API call to install the boost. Also, verify that the boost column changes to the progress icon and you receive a success notification.

9. Click the info icon next to the column names and verify the modal opens up as expected, and that the buttons work as expected.

<img width="1070" alt="Screenshot 2023-09-26 at 4 13 46 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/24ed55d3-904f-4eb4-a807-b7cfdec5504a">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yashwin committed:
Update LicenseLightbox to render the extraContent

yashwin committed:
Create a new component to display the license info modal for reusability

yashwin committed:
Create use-request-boost-score-mutation.ts

yashwin committed:
Create use-install-plugin-mutation.ts

yashwin committed:
Create use-install-boost.tsx

yashwin committed:
Create BoostLicenseInfoModal and use it to show boost license info modal

yashwin committed:
Fix test cases

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-content/index.tsx:
```diff
@@ -136,15 +97,7 @@ const SiteContent = ( { data, isLoading, currentPage, isFavoritesTab }: Props, r
-					onClose={ onHideLicenseInfo }
-				/>
-			) }
+			{ currentLicenseInfo && <SiteBoostLicenseModal currentLicenseInfo={ currentLicenseInfo } /> }
```
I would suggest naming this `LicenseInfoModal` so it doesn't make an expectation that this modal is specific to `Boost`.

yashwin replied:
Good catch.

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/license-info-modal/index.tsx:
```diff
@@ -0,0 +1,86 @@
+	onCtaClick?: () => void;
+}
+
+export default function LicenseInfoModal( {
```
👍 thanks for moving this to a separate component.

mattgawarecki commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/hooks/use-install-boost.tsx:
```diff
@@ -0,0 +1,111 @@
+		await queryClient.cancelQueries( queryKey );
+
+		// Update to the new value
+		const updatedData = queryClient.setQueryData( queryKey, ( oldSites: any ) => {
+			return {
+				...oldSites,
+				sites: oldSites?.sites.map( ( site: Site ) => {
+					if ( site.blog_id === siteId ) {
+						return {
+							...site,
+							has_pending_boost_one_time_score: true,
+						};
+					}
+					return site;
+				} ),
+			};
+		} );
+		queryClient.setQueryData( queryKey, updatedData );
```
Are we making two requests here? I see `setQueryData` called on line 42, and again on line 56. I'm not familiar with `react-query` though, so if I'm misreading, let me know 🙂

yashwin replied:
Good catch 👁️

mattgawarecki commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/license-info-modal/index.tsx:
```diff
@@ -0,0 +1,86 @@
+	};
+
+	const onIssueLicense = () => {
+		if ( currentLicenseProductSlug ) {
+			recordEvent( 'issue_license_info', {
+				product: currentLicenseProductSlug,
+			} );
+			onCtaClick?.();
+			onHideLicenseInfo();
+			page(
+				addQueryArgs(
+					{
+						product_slug: currentLicenseProductSlug,
+						source: 'dashboard',
+						site_id: siteId,
+					},
+					'/partner-portal/issue-license/'
+				)
+			);
+		}
```
Non-blocking: If you like the idea, we can reduce nesting here by negating the if condition --

```suggestion
		if ( ! currentLicenseProductSlug ) {
			return;
		}
		
		recordEvent( 'issue_license_info', {
			product: currentLicenseProductSlug,
		} );
		onCtaClick?.();
		onHideLicenseInfo();
		page(
			addQueryArgs(
				{
					product_slug: currentLicenseProductSlug,
					source: 'dashboard',
					site_id: siteId,
				},
				'/partner-portal/issue-license/'
			)
		);
```

yashwin replied:
Done

mattgawarecki commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-expanded-content/boost-site-performance.tsx:
```diff
@@ -44,7 +48,7 @@ export default function BoostSitePerformance( {
 	);
 
-	const isEnabled = !! ( hasBoost || overallScore );
+	const isEnabled = !! ( hasBoost || overallScore ) || hasPendingScore;
```
Non-blocking: it looks like maybe the parentheses here aren't necessary, except for possibly a TypeScript complaint. An alternative might look something like this --

```suggestion
	const isEnabled = Boolean( hasBoost || overallScore || hasPendingScore );
```

Or, since the only non-Boolean value is `overallScore` --

```suggestion
	const isEnabled = hasBoost || Boolean( overallScore ) || hasPendingScore;
```

yashwin replied:
Agree!

yashwin committed:
Address PR comments

yashwin committed:
Remove useRequestBoostScore as it is no longer required

yashwin commented:
@jkguidaven @mattgawarecki Thanks for your thorough and quick review. I have addressed all the comments. Could you please take another look?

Changes done:

- Addressed PR comments: https://github.com/Automattic/wp-calypso/pull/82239/commits/e34d726be5a855bfc179551ff8aa34b83454a720
- We removed the code behind making the API call request to request the boost score as it is no longer required: https://github.com/Automattic/wp-calypso/pull/82239/commits/bca43846441bb60a7fd4b0287ad2b276ecd4c03b. I have also updated the testing instructions accordingly. 

There are some more changes coming this. But that will be mainly design changes.

yashwin committed:
Add notice with a link to Jetpack Boost

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-expanded-content/boost-site-performance.tsx:
```diff
@@ -117,10 +121,10 @@ export default function BoostSitePerformance( {
-							<InProgressIcon />
-						) : (
-							<>
+						<>
```
It's non-blocking, but we could remove this so it doesn't add additional indention.

jkguidaven commented on client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-expanded-content/index.tsx:
```diff
@@ -50,6 +50,7 @@ export default function SiteExpandedContent( {
 					boostData={ boostData }
 					siteId={ site.blog_id }
 					siteUrlWithScheme={ siteUrlWithScheme }
+					siteUrl={ site.url }
```
For consistency, similar to how we did for the rest of the columns, we can pass the site object as props instead of having multiple props (e.g., siteId, site. URL, etc....). wdyt?

jkguidaven commented on client/jetpack-cloud/sections/partner-portal/license-lightbox/index.tsx:
```diff
@@ -20,6 +20,7 @@ export type LicenseLightBoxProps = {
 	onActivate: ( product: APIProductFamilyProduct ) => void;
 	onClose: () => void;
 	product: APIProductFamilyProduct;
+	extraContent?: JSX.Element;
```
To avoid confusion that we are adding extra content in the sidebar and not in the main container, maybe we can name this `extraAsideContent` and make this more specific to the aside content. This way, the intention is more apparent.

yashwin replied:
Makes sense!

jkguidaven reviewed (commented):
Thanks for addressing my initial PR comment. I have tested it, and it is working as expected. I have left a few comments and suggestions.

yashwin committed:
Address PR comments

yashwin commented:
Thanks for the review again, @jkguidaven. I have addressed all your comments and made the changes: https://github.com/Automattic/wp-calypso/pull/82239/commits/dc3d346ec4eac0d0476c2410b76983660984220b

jkguidaven reviewed (approved):
Thanks for addressing my comments. I don't have anything else. LGTM 👍

mattgawarecki reviewed (approved):
I re-reviewed both the code and via functional testing. Everything behaved as expected, and the changes make sense to me. Adding my approval 👍

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9090700

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9090700).
